File: backend/routers/transactions.py
from datetime import date as date_type

from fastapi import APIRouter

from controllers import transactions as transactions_controller

router = APIRouter()


@router.get("/transactions")
async def get_transactions(
    category: str | None = None,
    status: str | None = None,
    date_from: date_type | None = None,
    date_to: date_type | None = None,
    amount_min: float | None = None,
    amount_max: float | None = None,
    search: str | None = None,
    sort_by: str = "date",
    sort_dir: str = "desc",
    page: int = 1,
    limit: int = 50,
    include_summary: bool = False,
):
    rows_payload, summary = await transactions_controller.fetch_transactions(
        category,
        status,
        date_from,
        date_to,
        amount_min,
        amount_max,
        search,
        sort_by,
        sort_dir,
        page,
        limit,
        include_summary,
    )
    return {"transactions": rows_payload, "summary": summary}

File: client/src/api/transactions.ts
import { apiGet } from "@/lib/api";

export type Transaction = {
  id: number;
  merchant: string;
  amount: number;
  date: string;
  category: string;
  status: string;
  description: string | null;
};

export type TransactionsQuery = {
  category?: string;
  status?: string;
  date_from?: string;
  date_to?: string;
  amount_min?: number;
  amount_max?: number;
  search?: string;
  sort_by?: "date" | "amount";
  sort_dir?: "asc" | "desc";
  page?: number;
  limit?: number;
  include_summary?: boolean;
};

export type TransactionsPage = {
  rows: Transaction[];
  page: number;
  limit: number;
  total_count: number;
};

export type CategorySummary = {
  category: string;
  total: number;
  count: number;
};

export type MonthSummary = {
  month: string;
  total: number;
};

export type TransactionsSummary = {
  by_category: CategorySummary[];
  by_month: MonthSummary[];
} | null;

export type TransactionsResponse = {
  transactions: TransactionsPage;
  summary: TransactionsSummary;
};

export function fetchTransactions(query: TransactionsQuery = {}) {
  return apiGet<TransactionsResponse>("/transactions", query);
}

export type TransactionFilterOptions = {
  categories: string[];
  statuses: string[];
};

export function fetchFilterOptions() {
  return apiGet<TransactionFilterOptions>("/transactions/filters");
}
